### ISSUE TITLE
Add SUPIR Upscaler

### ISSUE THREAD

**Issue by DN6** [huggingface/diffusers#7219]
### Model/Pipeline/Scheduler description

[SUPIR](https://supir.xpixel.group/) is a super-resolution model that looks like it produces excellent results

Github Repo: https://github.com/Fanghua-Yu/SUPIR

The model is quite memory intensive, so the optimisation features available in diffusers might be quite helpful in making this accessible to lower resource GPUs.    

### Open source status

- [X] The model implementation is available.
- [X] The model weights are available (Only relevant if addition is not a scheduler).

**Comment by ihkap11** [comment]
Hey can I work on this?

**Comment by yiyixuxu** [comment]
@ihkap11 hey! sure!

**Comment by Bhavay-2001** [comment]
Hi @yiyixuxu, anyone working on this? Can I also contribute? Please let me know how may I proceed?

**Comment by ihkap11** [comment]
Hey @Bhavay-2001 I'm currently working on this. Will post the PR here soon. 
I can tag you on the PR if I there is something I need help with :)

**Comment by Bhavay-2001** [comment]
ok great. Pls let me know.
Thanks

**Comment by landmann** [comment]
@ihkap11 how's it going 😁 I'd loooooove to have this

**Comment by ihkap11** [comment]
Hey I'll post the PR this weekend. :) apologies for delay

**Comment by ihkap11** [comment]
Initial Update:
- Understood the paper (paper highlights below)
- Currently defining paper components that will become diffuser artefacts. (WIP: breaking down SUPIR code)

Todos:
- 

<details>
   <summary><strong> Paper Insights</strong></summary>
<br>

### Motivation:

-  IR methods based on generative priors leverage powerful pre-trained generative models to introduce high-quality generation and prior knowledge into IR, bringing significant progress in
perceptual effects and intelligence of IR results.
- Continuously enhancing the capabilities of the generative prior is key to achieving more intelligent IR results, with model scaling being a crucial and effective approach.
- The authors propose scaling up generative priors and training data to address these limitations.


### Architecture Overview:
1. **Generative Prior**: The authors choose SDXL (Stable Diffusion XL) as the backbone for their generative prior due to its high-resolution image generation capability without hierarchical design.

2.  **Degradation-Robust Encoder**: They fine-tune the SDXL encoder to make it robust to degradation, enabling effective mapping of low-quality (LQ) images to the latent space.

3. **Large-Scale Adaptor**:  The author designed a new adaptor with network trimming and a ZeroSFT connector to control the generation process at the pixel level.
     <details>
     <summary><strong>Issues with existing adaptors</strong></summary>
     <br>

     - LoRA limits generation but struggles with LQ image control
     - T2I lacks the capacity for effective LQ image content identification
     - ControlNet’s direct copy is challenging for the SDXL model scale.
     </details>

    1. **Network Trimming**:  Modify the adaptor architecture by trimming half of the ViT blocks in each encoder block (of SDXL) to achieve a balance between network capacity and computational feasibility.
    2.  **Redesigning the Connector**: The introduced ZeroSFT module is built upon zero convolution and incorporates an additional spatial feature transfer (SFT) operation and group normalization.  
     

      <details>
      <summary><strong>Why do we need this?</strong></summary>
      <br>

      - The authors note that while SDXL's generative capacity delivers excellent visual effects, it also makes precise pixel-level control challenging.
      - ControlNet uses zero convolution for generation guidance, but relying solely on residuals is insufficient for the level of control required by IR tasks.
       </details> 

4. **Multi-Modality Language Guidance**: They incorporate the LLaVA multi-modal large language model to understand image content and guide the restoration process using textual prompts.

5. **Restoration-Guided Sampling**: They propose a modified sampling method to selectively guide the prediction results to be close to the LQ image, ensuring fidelity in the restored image.

</details>

### Thoughts on implementation details:
- Trainable components are degradation robust encoder and trimmed ControlNet.
- Extend the SDXL class from Diffusers and use SDXL checkpoint = `sd_xl_base_1.0_0.9vae.safetensors` as base pre-trained generative prior.
- The SUPIR model will first load pre-trained weights from the SDXL checkpoint, then it will load SUPIR-specific weights, which include the modifications and additions made to adapt the SDXL model for image restoration tasks.
- Trimmed ControlNet encoder which trims half of the ViT blocks from each encoder block. (Todo: Figure out where to make this change)
-  In the SUPIR model, SDXL (Stable Diffusion XL) is used as the backbone for the generative prior, and the [GLVControl ](https://github.com/Fanghua-Yu/SUPIR/blob/358a7a3cae0dada4830be993fc7229a2adfe8cc4/SUPIR/modules/SUPIR_v0.py#L155)and [LightGLVUNet](https://github.com/Fanghua-Yu/SUPIR/blob/358a7a3cae0dada4830be993fc7229a2adfe8cc4/SUPIR/modules/SUPIR_v0.py#L543C7-L543C19) modules are used as the adaptor to guide the SDXL model for image restoration. _Todo: Convert to Diffusers Artifact_
- Probably, a dummy code would look like this:
```
class SUPIRModel(nn.Module):
    def __init__(self, sdxl_model_path):
        super().__init__()
        self.sdxl_pipeline = StableDiffusionXLPipeline.from_pretrained(sdxl_model_path)
        self.glv_control = GLVControl(in_channels=3, out_channels=64, context_dim=128)
        self.light_glv_unet = LightGLVUNet(in_channels=3, out_channels=3)
        
    def forward(self, lq_image, context, num_inference_steps=50):
        # Generate control signal using GLVControl
        control_signal = self.glv_control(lq_image, context)
        
        # Use SDXL pipeline for guided diffusion
        restored_image = self.sdxl_pipeline(
            prompt="",
            image=lq_image,
            control_image=control_signal,
            num_inference_steps=num_inference_steps,
            generator=None,
        ).images[0]
        
        # Refine the restored image using LightGLVUNet
        refined_image = self.light_glv_unet(restored_image, control_signal)
        
        return refined_image
```
- [ZeroFST](https://github.com/Fanghua-Yu/SUPIR/blob/358a7a3cae0dada4830be993fc7229a2adfe8cc4/SUPIR/modules/SUPIR_v0.py#L62) acts as a connector. _Todo: Convert to Diffusers Artifact_

To cover later: 
- LLaVA for multi-modality language guidance.


I'm currently in the process of breaking down SUPIR code into diffusers artefacts and figuring out optimization techniques to make it compatible with low-resource GPUs. 





Feel free to correct me or start a discussion on this thread. Let me know if you wish to collaborate, I'm happy to set up discussions and work on it together :).

**Comment by landmann** [comment]
Looks fantastic! How far along did you get, @ihkap11 ?

Btw, a good reference for the input parameters are here https://replicate.com/cjwbw/supir?prediction=32glqstbvpjjppxmvcge5gsncu

**Comment by landmann** [comment]
@ihkap11 how you doing? Which part are you stuck?

**Comment by ihkap11** [comment]
Hey @landmann, I'm finding it hard to map a few components from the paper's network architecture details to the codebase they provided. 

Currently, I'm stuck on understanding how they are trimming the VIT blocks when using the modified ControlNet adapter with [ZeroFST](https://github.com/Fanghua-Yu/SUPIR/blob/358a7a3cae0dada4830be993fc7229a2adfe8cc4/SUPIR/modules/SUPIR_v0.py#L62) connector at the code level.  They seem to use [GLVControl](https://github.com/Fanghua-Yu/SUPIR/blob/358a7a3cae0dada4830be993fc7229a2adfe8cc4/SUPIR/modules/SUPIR_v0.py#L155) but no VIT component and network trimming that I can spot in the codebase.

I sent an email to one of the authors. If I don't hear back, I plan to follow up with more specific questions this week.

If interested would you want to take a second look and share your thoughts?

**Comment by landmann** [comment]
@austinfujimori you can take a look if you're free 🙂

**Comment by CuddleSabe** [comment]
> @ihkap11 are you following the paper and trying to code it? Why not just make a wrapper around what they have? It's pytorch after all, no? I haven't read the paper much in depth, but was able to run [SUPIR](https://github.com/Fanghua-Yu/SUPIR) locally!
> 
> @austinfujimori you should take a look if you're free 🙂

the "ZeroSFT" is to replace the  concat[hidden_states, res_sample] in AttnUpBlock and CrossAttnUpBlock, so we can't use diffusers's sdxl pipeline to implement it.

**Comment by CuddleSabe** [comment]
> > @ihkap11 are you following the paper and trying to code it? Why not just make a wrapper around what they have? It's pytorch after all, no? I haven't read the paper much in depth, but was able to run [SUPIR](https://github.com/Fanghua-Yu/SUPIR) locally!
> > @austinfujimori you should take a look if you're free 🙂
> 
> the "ZeroSFT" is to replace the concat[hidden_states, res_sample] in AttnUpBlock and CrossAttnUpBlock, so we can't use diffusers's sdxl pipeline to implement it.

because the different between the sgm and diffusers arch, its difficult

**Comment by ihkap11** [comment]
@CuddleSabe would you like to connect on discord to dicuss SUPIR and possibly collaborate in figuring out how to support it in diffusers?

**Comment by CuddleSabe** [comment]
> @CuddleSabe would you like to connect on discord to discuss SUPIR and possibly collaborate in figuring out how to support it in diffusers? You can find me by the username tortillachips11.

in sgm, it like this
<img width="398" alt="截屏2024-04-10 15 34 02" 
src="https://github.com/huggingface/diffusers/assets/61224076/fb622363-7940-4f5c-91ed-29cf9072f21a">

the "hs" is the res_sample from the down_blocks.
https://github.com/Stability-AI/generative-models/blob/fbdc58cab9f4ee2be7a5e1f2e2787ecd9311942f/sgm/modules/diffusionmodules/openaimodel.py#L849



but in diffusers, it like this
the "res_sample" in diffusers equal the "hs" in sgm
![截屏2024-04-10 15 36 52](https://github.com/huggingface/diffusers/assets/61224076/ae8a8380-18cc-4b0a-ac49-873cb2cfadf2)

https://github.com/huggingface/diffusers/blob/66f94eaa0c68a893b2aba1ec9f79ee7890786fba/src/diffusers/models/unets/unet_2d_condition.py#L1189


so you need to rewrite the AttnUpBlock2D 、CrossAttnUpBlock2D and UNetMidBlock2DCrossAttn in diffusers

**Comment by elismasilva** [comment]
hi @ihkap11, have news?

**Comment by ihkap11** [comment]
Hey! I tried but couldn't get this working. Feel free to take over the implementation for  this Issue.

**Comment by elismasilva** [comment]
> Hey! I tried but couldn't get this working. Feel free to take over the implementation for this Issue.

But do you have a branch where we can continue where you left off? I might try this after I finish a project I'm involved with.

**Comment by sayakpaul** [comment]
Cc: @asomoza

**Comment by asomoza** [comment]
Just in case, this is not an easy task, everything is in the sgm format so there's a lot of conversion involved. It requires a deep understanding of the original code and diffusers.

Probably the best choice here is to start as a research project and convert all the sgm code to diffusers, and then when stuck, get help from the maintainers and the community.

**Comment by tolgacangoz** [comment]
Can I work on this research project?

**Comment by asomoza** [comment]
sure, we still want to be able to use it with diffusers.

**Comment by zdxpan** [comment]
accoding to the paper and the Comfyui 
will impliment below:

<img width="889" alt="image" src="https://github.com/user-attachments/assets/191cce35-774b-4be8-ae5e-02d3bb04184d">
SUPIR MODEL_LOADER -> SUPIR_MODEL, SUPIR_vae
<img width="325" alt="image" src="https://github.com/user-attachments/assets/54889ec2-054c-4337-8c79-43e8023b1b79">

SUPIR_FIRSTSTAGE Denoiser     : take Low quality image in and blur or smooth image out and it`s latent
- this stage include an SUPIR VAE  include vae-encoder vae-decoder

<img width="532" alt="image" src="https://github.com/user-attachments/assets/0f517280-a8e5-453f-8cf8-dd7976e6b6ee">
SUPIR-controlnet  : which take latents in and time stpes in , generate controlnet residuals dowsamples and midsample out
 
<img width="456" alt="image" src="https://github.com/user-attachments/assets/30475f7b-0df3-4b0e-a027-f708b7d9a236">
An hacked Unet which modify the connector of each dow and up blocks  use ZeroSFT

**Comment by elismasilva** [comment]
hi @DN6 i already did it and it is working fine see results: https://imgsli.com/Mzg2NTUx/0/1. But then I saw that they changed the license and it seems that it is very restrictive. I don't know how to deal with this case. I even sent an email to the owner but it's been a month since I got a response.

**Comment by asomoza** [comment]
They changed the license from MIT to "something unnamed" literally two days after the original post here, didn't see that.

Don't know why they keep changing it to a custom one, probably because there's some other party involved, but basically, this license is the same as Flux, you can use it and change it freely if it's non-commercial which is fine, we have other models with the same type of custom license.

AFAIK this is still the SOTA upscaler model so it would be nice to have it.

**Comment by elismasilva** [comment]
> They changed the license from MIT to "something unnamed" literally two days after the original post here, didn't see that.
> 
> Don't know why they keep changing it to a custom one, probably because there's some other party involved, but basically, this license is the same as Flux, you can use it and change it freely if it's non-commercial which is fine, we have other models with the same type of custom license.
> 
> AFAIK this is still the SOTA upscaler model so it would be nice to have it.

they have this on their new license: "_b. Proprietary Neural Network Components: Notwithstanding the open-source license granted in Section 2(a) for open-source components, the Licensor retains all rights, title, and interest in and to the proprietary Neural Network Components developed and trained by the Licensor. The license granted herein **does not include rights** to use, copy, modify, distribute, or create derivative works of these proprietary Neural Network Components without obtaining express written permission from the Licensor_." 

The license is a bit contradictory because at the same time it says that parts of the code are open source, another part cannot be changed. I wonder if in the case where I converted the model to .safetensors I would be violating this clause.

**Comment by asomoza** [comment]
I understand that part (I'm not expert) as to referring to derivative works only, that's a clause for limiting the use of the model to do something similar to circumvent the license and then use it commercially, like using its outputs to train a similar model.

When integrating it with diffusers and changing the format for diffusers, and of course copying the license and referencing them as copyright owners, you're not using it as a derivative work but as the original model.

On the other part, if we interpret it as for everything and not just derivative works, it invalidates all from before and at that point they should just remove "open source" from it and just say that any use of the model will need their permission.

But still it would be a lot better to get the permission directly from the authors, since it's a not easy to understand license.

**Comment by elismasilva** [comment]
I agree, it's not easy to understand this license. Let's see if anyone else has any thoughts on this. I don't think the authors will answer me as it's been a long time.

**Comment by vladmandic** [comment]
imo, it's a non-standard license, but it's pretty clear since it several times explicitly defines " neural network architecture, weights and biases", not model file and/or packaging as such. conversion to safetensors is NOT derivative work since it fully maintains original "architecture, weights and biases".

as an example, under this license, converting to something like gguf with pre-quantized weights would be a no-no without explicit permission.

**Comment by elismasilva** [comment]
> imo, it's a non-standard license, but it's pretty clear since it several times explicitly defines " neural network architecture, weights and biases", not model file and/or packaging as such. conversion to safetensors is NOT derivative work since it fully maintains original "architecture, weights and biases".
> 
> as an example, under this license, converting to something like gguf with pre-quantized weights would be a no-no without explicit permission.

Good point. And in this case I saw a version of the model files that seems to have been pruned because it was only 2GB or they converted it to fp16 I don't know for sure. In my case I just split the modules in diffusers format to be loaded as pretrained so at most what happened is that some keys had to be adapted in the case of the encoder, even so there was no change in the content.

**Comment by OlegRuban-ai** [comment]
Could i use SUPIR in duffusers now? I dodn't find correct info for integration this model

**Comment by asomoza** [comment]
we don't have the integration yet, so no, you can't use SUPIR with diffusers right now.

**Comment by elismasilva** [comment]
Hi everyone, the first stage of the process is complete, you can use and test it. The pipeline is already functional and I believe that even if it's not included in the core, it will still be useful for the community.

Try Demo here: https://huggingface.co/spaces/elismasilva/sup-toolbox-app

App repository: https://github.com/DEVAIEXP/sup-toolbox-app

CLI repository: https://github.com/DEVAIEXP/sup-toolbox
